src/msgpack.lua

```diff
@@ -0,0 +1,220 @@
+                          jit.arch == 'ppc' or
+                          jit.arch == 'mips')
+
+if strict_alignment then
+   intbuff = ffi.new('union intbuff[1]')
+end
+
+local function bswap_u16(num)
+    return bit.rshift(bit.bswap(tonumber(num)), 16)
+end
+
+local function decode_u16_risc(data)
+    ffi.copy(intbuff, data[0], 2)
+    data[0] = data[0] + 2
+    return tonumber(bswap_u16(intbuff[0].u16))
+end
+
+local function decode_u32_risc(data)
+    ffi.copy(intbuff, data[0], 4)
+    data[0] = data[0] + 4
+    return tonumber(ffi.cast('uint32_t', bit.bswap(tonumber(intbuff[0].u32))))
+end
+
+local function decode_u16_cisc(data)
+    local num = bswap_u16(ffi.cast(uint16_ptr_t, data[0])[0])
+    data[0] = data[0] + 2
+    return tonumber(num)
+end
+
+local function decode_u32_cisc(data)
+    local num = ffi.cast('uint32_t', bit.bswap(tonumber(ffi.cast(uint32_ptr_t, data[0])[0])))
+    data[0] = data[0] + 4
+    return tonumber(num)
+end
+
+local decode_u16
+local decode_u32
+
+if strict_alignment then
+    decode_u16 = decode_u16_risc
+    decode_u32 = decode_u32_risc
+else
+    decode_u16 = decode_u16_cisc
+    decode_u32 = decode_u32_cisc
+end
+
+local function decode_array_header_slowly(c, data)
+    -- fixarray: 0x90 .. 0x9f
+    if bit.band(c, 0xf0) == 0x90 then
+        return bit.band(c, 0xf)
+    end
+
+    -- array 16
+    if c == 0xdc then
+        return decode_u16(data)
+    -- array 32
+    elseif c == 0xdd then
+        return decode_u32(data)
+    else
+        error('Unsupported MP_ARRAY')
+    end
+end
+
+local function decode_array_header_compat(buf, size)
+    if type(buf) == "string" then
+        bufp[0] = ffi.cast(char_ptr_t, buf)
+    elseif ffi.istype(char_ptr_t, buf) then
+        bufp[0] = buf
+    end
+
+    local c = bufp[0][0]
+    bufp[0] = bufp[0] + 1
+    if (mp_typeof(c) ~= native.MP_ARRAY) then
+        error(string.format("%s: unexpected msgpack type '0x%x'",
+                            'decode_array_header_compat', c))
+    end
+
+    local len = decode_array_header_slowly(c, bufp)
+
+    return len, ffi.cast(ffi.typeof(buf), bufp[0])
+end
+
+local function decode_map_header_slowly(c, data)
+    -- fixmap: 0x80 .. 0x8f
+    if bit.band(c, 0xf0) == 0x80 then
+        return bit.band(c, 0xf)
+    end
+
+    -- map 16
+    if c == 0xde then
+        return decode_u16(data)
+    -- map 32
+    elseif c == 0xdf then
+        return decode_u32(data)
+    else
+        error('Unsupported MP_MAP')
+    end
+end
+
+local function decode_map_header_compat(buf, size)
+    bufp[0] = buf
+
+    local c = bufp[0][0]
+    bufp[0] = bufp[0] + 1
+    if (mp_typeof(c) ~= native.MP_MAP) then
+        error(string.format("%s: unexpected msgpack type '0x%x'",
+                            'decode_map_header_compat', c))
+    end
+
+    local len = decode_map_header_slowly(c, bufp)
+
+    return len, ffi.cast(ffi.typeof(buf), bufp[0])
+end
+
+local IPROTO_DATA_KEY = 0x30
+
+local function skip_request_header(self, buf)
```

Comment:
I don't like idea to pass "self" directly into module functions. Also if you want to use asserts in this code it's better to add some verbose error messages

Reply:
I don't entirely get your reasoning about not passing self to module methods, could you please clarify?

(Here I needed to have access to the context we created during module initialization, and it appears to be obvious that I needed to pass object to this context. Why should I avoid it?)

Reply:
I believe libraries should be stateless. Also you don't really need this state.

It's a just flag that could be easily replaced with
```lua
local skip_request_header = function() end -- default behaviour (1.10 <)
if msgpack.decode_map_header == nil then -- (1.10 >=)
    skip_request_header = function(buf)
        local len, key
        len, buf.rpos = self.decode_map_header(buf.rpos, buf:size())
        assert(len == 1)
        key, buf.rpos = self.decode_unchecked(buf.rpos)
        assert(key == IPROTO_DATA_KEY)
    end
end
```

At first, you eliminate "if fix_compat" check. Secondly, you don't confuses users with magic "library" context.
As usually we use `library.call(...)` and `some_class_instance:call(...)`.
Finally, you don't really need to expose `fix_compat` flag.
You can just define local variable and use it

```lua
local fix_compat = msgpack.decode_map_header == nil
local function skip_request_header(buf)
    if not fix_compat then
        return
    end
    ....
end
```